File: packages/ocp-shuffle/src/rng.ts
import { hashToScalar, scalarToBytes, u32le } from "@onchainpoker/ocp-crypto";
import type { Scalar } from "@onchainpoker/ocp-crypto";

export interface ScalarRng {
  nextScalar(): Scalar;
}

export class DeterministicRng implements ScalarRng {
  private readonly seed: Uint8Array;
  private counter = 0;

  constructor(seed: Uint8Array) {
    if (!(seed instanceof Uint8Array) || seed.length === 0) throw new Error("DeterministicRng: empty seed");
    this.seed = seed;
  }

  nextScalar(): Scalar {
    const c = u32le(this.counter++);
    return hashToScalar("ocp/v1/shuffle/rng", this.seed, c);
  }

  nextBytes(len: number): Uint8Array {
    if (!Number.isInteger(len) || len < 0) throw new Error("DeterministicRng.nextBytes: invalid length");
    const out = new Uint8Array(len);
    let off = 0;
    while (off < len) {
      const b = scalarToBytes(this.nextScalar());
      const take = Math.min(b.length, len - off);
      out.set(b.subarray(0, take), off);
      off += take;
    }
    return out;
  }
}

File: apps/chain/internal/ocpshuffle/rng.go
package ocpshuffle

import (
	"fmt"

	"onchainpoker/apps/chain/internal/ocpcrypto"
)

type scalarRng interface {
	NextScalar() (ocpcrypto.Scalar, error)
	NextBytes(n int) ([]byte, error)
}

type DeterministicRng struct {
	seed    []byte
	counter uint32
}

func NewDeterministicRng(seed []byte) (*DeterministicRng, error) {
	if len(seed) == 0 {
		return nil, fmt.Errorf("DeterministicRng: empty seed")
	}
	return &DeterministicRng{seed: append([]byte(nil), seed...)}, nil
}

func (r *DeterministicRng) NextScalar() (ocpcrypto.Scalar, error) {
	c := make([]byte, 4)
	c[0] = byte(r.counter)
	c[1] = byte(r.counter >> 8)
	c[2] = byte(r.counter >> 16)
	c[3] = byte(r.counter >> 24)
	r.counter++
	return ocpcrypto.HashToScalar("ocp/v1/shuffle/rng", r.seed, c)
}

func (r *DeterministicRng) NextBytes(n int) ([]byte, error) {
	if n < 0 {
		return nil, fmt.Errorf("DeterministicRng.NextBytes: invalid length")
	}
	out := make([]byte, n)
	off := 0
	for off < n {
		s, err := r.NextScalar()
		if err != nil {
			return nil, err
		}
		sb := s.Bytes()
		take := len(sb)
		if n-off < take {
			take = n - off
		}
		copy(out[off:], sb[:take])
		off += take
	}
	return out, nil
}

